### DISCUSSION TITLE
Is RDCYCLE available?

### DISCUSSION THREAD

**Discussion by iagocaran** [gem5/gem5#1397] [Q&A] [answered]
I am adapting some benchmark code to run on gem5 using the RISC-V and it uses the `rdcycle` pseudo instruction to get the cycle count that is used as a time metric. Using the `simple-riscv.py` configuration as an example, the `rdcycle` instruction always returns zero, the same happens if I change it to use the **TimingSimpleCPU**.
Are there any specific configurations to use `rdcycle` with gem5 or is it not implemented?

Below is the code snippet that the benchmark uses.

`unsigned long read_cycles(void)
{
    unsigned long cycles;
    asm volatile ("rdcycle %0" : "=r" (cycles));
}`

**Comment by robhau** [comment, marked as answer]
Hi @iagocaran,

I think you also have to return `cycles` inside your function snippet.

If you take a look at the [asm code of your function](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:49,endLineNumber:3,positionColumn:49,positionLineNumber:3,selectionStartColumn:49,selectionStartLineNumber:3,startColumn:49,startLineNumber:3),source:'unsigned+long+read_cycles(void)+%7B%0A++++unsigned+long+cycles%3B%0A++++asm+volatile+(%22rdcycle+%250%22+:+%22%3Dr%22+(cycles))%3B%0A%7D'),l:'5',n:'1',o:'C+source+%231',t:'0')),k:31.751167313549185,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:rv64-cgcc1410,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-Os',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+RISC-V+(64-bits)+gcc+14.1.0+(Editor+%231)',t:'0')),k:26.22411226267211,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+clang+16.0.0',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+RISC-V+(64-bits)+gcc+14.1.0+(Compiler+%231)',t:'0')),k:42.02472042377869,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4), `cycles` is not necessarily stored in `a0`. However, `a0` and `a1` are the return register in RISC-V, so your result is undefined.

If add a return statement in your code

```C
unsigned long read_cycles(void) {
    unsigned long cycles;
    asm volatile ("rdcycle %0" : "=r" (cycles));
    return cycles;
}
```

[`cycles` is stored in `a0 ` instead](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:19,endLineNumber:4,positionColumn:19,positionLineNumber:4,selectionStartColumn:19,selectionStartLineNumber:4,startColumn:19,startLineNumber:4),source:'unsigned+long+read_cycles(void)+%7B%0A++++unsigned+long+cycles%3B%0A++++asm+volatile+(%22rdcycle+%250%22+:+%22%3Dr%22+(cycles))%3B%0A++++return+cycles%3B%0A%7D'),l:'5',n:'1',o:'C+source+%231',t:'0')),k:31.751167313549185,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:rv64-cgcc1410,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-Os',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+RISC-V+(64-bits)+gcc+14.1.0+(Editor+%231)',t:'0')),k:26.22411226267211,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+clang+16.0.0',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+RISC-V+(64-bits)+gcc+14.1.0+(Compiler+%231)',t:'0')),k:42.02472042377869,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).

**Reply by iagocaran** [reply]
Hi @robhau,

That was a copy and paste mistake, even after adding the line it did not work.
After further investigation, I found some problems elsewhere.

Everything is working now, thanks.

**Reply by robhau** [reply]
Hi @iagocaran,

thank you for your answer. So the problem was not gem5 related, right? On my side, the pseudo instruction works.
